multitenant: account for multi-region transfers in RU accounting model

jaylim-crl opened the pull request:
#### tenantcostmodel: add RU multiplier for read and write requests 

Previously, we multiplied consumed RU for write requests with the number of
replicas (i.e. replication factor). This commit decouples that, and introduces
new writeMultiplier (and readMultiplier) variables that explicitly denote
the multipliers for the write (and read) requests. For now, we will set
writeMultiplier=numReplicas and readMultiplier=1, and it is clear that this
model does not account for cross-region transfers since we are assuming that
all write requests (including replication) consume the same amount of RUs.
A future commit will address that.

This commit should be purely mechanical, and is a no-op.

#### multitenant: account for multi-region transfers in RU accounting model

Related to https://cockroachlabs.atlassian.net/browse/CRDB-19074.

The existing RU model assumes a fixed cost per GB transferred over the network.
By default, the cost used is the average cost per GB when 1 replica is in the
same AZ, and 2 replicas are in the other AZs. In multi-region clusters, there
are cross-region transfers that do not fit well with that model.

This commit addresses that issue by updating the RU accounting model to include
multi-region transfers. We do so by introducing a regional cost multiplier
table which operators could set via a new cluster setting
(i.e. `tenant_cost_model.regional_cost_multiplier_table`). When set, all
cross-region transfers within the KV client will use that multiplier table by
ensuring that the final consumed RU is multiplied by some value in that table.
If a region-to-region pair is not found in that table, an RU multiplier of 1
is assumed. In this multiplier table model, it is assumed that the multiplier
value for regions are the same bidirectionally.

Here's an example of a valid cluster setting value for nodes with regional
localities of us-east1, eu-central1, and asia-southeast1.

```
{"regions":["us-east1","eu-central1","asia-southeast1"],"matrix":[[1,1.5,2.6],[1,3.5],[1]]}
```

The data above represents a table of:

```
                | us-east1 | eu-central1 | asia-southeast1
-----------------------------------------------------------
   us-east1     |    1     |     1.5     |       2.6
  eu-central1   |   1.5    |      1      |       3.5
asia-southeast1 |   2.6    |     3.5     |        1
```

Epic: [CRDB-19074](https://cockroachlabs.atlassian.net/browse/CRDB-19074)

Release note: None

jaylim-crl committed:
multitenant: account for multi-region transfers in RU accounting model

Related to https://cockroachlabs.atlassian.net/browse/CRDB-19074.

The existing RU model assumes a fixed cost per GB transferred over the network.
By default, the cost used is the average cost per GB when 1 replica is in the
same AZ, and 2 replicas are in the other AZs. In multi-region clusters, there
are cross-region transfers that do not fit well with that model.

This commit addresses that issue by updating the RU accounting model to include
multi-region transfers. We do so by introducing a regional cost multiplier
table which operators could set via a new cluster setting
(i.e. `tenant_cost_model.regional_cost_multiplier_table`). When set, all
cross-region transfers within the KV client will use that multiplier table by
ensuring that the final consumed RU is multiplied by some value in that table.
If a region-to-region pair is not found in that table, an RU multiplier of 1
is assumed. In this multiplier table model, it is assumed that the multiplier
value for regions are the same bidirectionally.

Here's an example of a valid cluster setting value for nodes with regional
localities of us-east1, eu-central1, and asia-southeast1.

```
{"regions":["us-east1","eu-central1","asia-southeast1"],"matrix":[[1,1.5,2.6],[1,3.5],[1]]}
```

The data above represents a table of:

```
                | us-east1 | eu-central1 | asia-southeast1
-----------------------------------------------------------
   us-east1     |    1     |     1.5     |       2.6
  eu-central1   |   1.5    |      1      |       3.5
asia-southeast1 |   2.6    |     3.5     |        1
```

Epic: CRDB-19074

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95913)
<!-- Reviewable:end -->

jaylim-crl commented:
The test failures are unrelated:
- TestLogic_unique/Upsert: https://github.com/cockroachdb/cockroach/pull/96001
- TestColdStartLatency: https://github.com/cockroachdb/cockroach/pull/96023

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, very nice, just some suggestions and nits from me.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-:-NMqqb-68INAn7OOQv2n:beeuoce)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jaylim-crl, @JeffSwenson, and @RaduBerinde)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2245 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMq7Bpt61VrCYE4nCu1:-NMq7Bpt61VrCYE4nCu2:b-3iv01f) ([raw file](https://github.com/cockroachdb/cockroach/blob/858dbeb96ff0b514bfc203d6c892a4bc8cec4b9c/pkg/kv/kvclient/kvcoord/dist_sender.go#L2245)):*
> ```Go
> 
> 				if ds.kvInterceptor != nil {
> 					var fromRegion, toRegion string
> ```

I think this chunk of logic belongs in a helper function. Especially once you add some extra logging/comments, it deserves it's own function - it might even deserve a little unit test.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2247 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMq6hMe7VHNW1CwXQwg:-NMq6hMe7VHNW1CwXQwh:biaqepf) ([raw file](https://github.com/cockroachdb/cockroach/blob/858dbeb96ff0b514bfc203d6c892a4bc8cec4b9c/pkg/kv/kvclient/kvcoord/dist_sender.go#L2247)):*
> ```Go
> 					var fromRegion, toRegion string
> 					if idx := replicas.Find(curReplica.ReplicaID); idx != -1 {
> 						fromRegion, _ = ds.locality.Find("region")
> ```

NIT: It's unfortunate that we hardcode a particular locality tier name here. Ideally, we'd have a cluster setting that specifies the name or some other way to configure it. That said, it's possible and maybe likely it'd ever be used for anything but Serverless, so not sure it's worth the trouble. At the very least, we should have a comment about it.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMq7LPv2WBuGR77cpsi:-NMq7LPv2WBuGR77cpsj:bdnmnc7) ([raw file](https://github.com/cockroachdb/cockroach/blob/858dbeb96ff0b514bfc203d6c892a4bc8cec4b9c/pkg/kv/kvclient/kvcoord/dist_sender.go#L2248)):*
> ```Go
> 					if idx := replicas.Find(curReplica.ReplicaID); idx != -1 {
> 						fromRegion, _ = ds.locality.Find("region")
> 						for _, tier := range replicas[idx].Tiers {
> ```

Similar to below, I think we need to log an error if we can't find "region" in the tiers.
___
*[`pkg/kv/kvserver/replica_rate_limit.go` line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMq7wpSCs-BrcP-A0tZ:-NMq7wpSCs-BrcP-A0t_:b-h02pmy) ([raw file](https://github.com/cockroachdb/cockroach/blob/858dbeb96ff0b514bfc203d6c892a4bc8cec4b9c/pkg/kv/kvserver/replica_rate_limit.go#L31)):*
> ```Go
> 		return nil
> 	}
> 	// Both fromRegion and toRegion are not needed here, and is only used to
> ```

NIT: Actually, this code is not used to calculate RU costs - it's used to estimate CPU usage on KV nodes. The better term is "KV Compute Units". See this comment in `tenantrate/settings.go`:
```
// We limit the rate of KV operations using the tenant cost model which can
// map these operations to "KV Compute Units". Note that KV Compute Units are
// not the same thing as Request Units. One KV Compute Unit is equivalent to one
// millisecond of CPU. Request Units measure other resources besides CPU (e.g.
// egress and IOPS). Also, Request Units correspond to the retail price of
// resources rather than the underlying cost of Cloud Provider resources (i.e.
// they include markup). For example, one KV Compute Unit might correspond to 2
// Request Units.
```

I'm noticing that some of the comments in the `tenantrate` package don't reflect this comment, so it's a bit confusing.
___
*[`pkg/multitenant/tenantcostmodel/model.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMq8sXi5knxPeBwm1Rw:-NMq8sXi5knxPeBwm1Rx:b16y211) ([raw file](https://github.com/cockroachdb/cockroach/blob/858dbeb96ff0b514bfc203d6c892a4bc8cec4b9c/pkg/multitenant/tenantcostmodel/model.go#L18)):*
> ```Go
> // into this abstract unit.
> //
> // To get an idea of the magnitude of an RU, the cost model was designed so that
> ```

Do you mind updating this comment to read:
```
To get an idea of the magnitude of an RU, the cost model was designed so that a small point read of < 64 bytes costs 1 RU.
```
___
*[`pkg/multitenant/tenantcostmodel/settings.go` line 278 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMqoUXJDfIAz9hYldLn:-NMqoUXKAvaAqhmZMeEs:bh7m7am) ([raw file](https://github.com/cockroachdb/cockroach/blob/858dbeb96ff0b514bfc203d6c892a4bc8cec4b9c/pkg/multitenant/tenantcostmodel/settings.go#L278)):*
> ```Go
> 	matrix := make(map[string]map[string]RUMultiplier)
> 	linkRegion := func(a, b string, val RUMultiplier) {
> 		if _, ok := matrix[a]; !ok {
> ```

How come we don't skip adding `1` multipliers into this table, since we always fall back to `1` during lookup if we can't find region(s)?
___
*[`pkg/multitenant/tenantcostmodel/settings.go` line 290 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMqnQo-8ngJD4vPsfBk:-NMqnQo-8ngJD4vPsfBl:b-6luc8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/858dbeb96ff0b514bfc203d6c892a4bc8cec4b9c/pkg/multitenant/tenantcostmodel/settings.go#L290)):*
> ```Go
> 			linkRegion(fromRegion, toRegion, multiplier)
> 			if fromIdx != toIdx {
> 				linkRegion(toRegion, fromRegion, multiplier)
> ```

NIT: Did you consider only putting the 1-way mappings in this table and checking both sides during lookup? It'd mean 2 lookups 50% of the time and 3 lookups 50% of the time, vs. 2 lookups 100% of the time. But it'd use 1/2 the memory. I think this way is fine, but just wanted to check that you'd made a conscious decision to go this way.


<!-- Sent from Reviewable.io -->

JeffSwenson commented on pkg/multitenant/tenantcostmodel/settings.go:
```diff
@@ -118,41 +134,261 @@ var (
+type RegionalCostMultiplierCompactTable struct {
+	// Regions represents the list of region names. The indexes of the regions
+	// will be used alongside with the matrix.
+	Regions []string `json:",omitempty"`
```
nit: I think we should use snake_case for the json keys.

CRDB isn't completely consistent with json style, but greping through the code base suggests that snake case is the dominant json naming convention.

The only other setting that accepts json is https://github.com/cockroachdb/cockroach/blob/fefa5fe16cc38c14a1406f6c91002493faa1cfb4/pkg/ccl/oidcccl/settings.go#L104 and it uses snake case as well.

jeffswenson reviewed (approved):
LGTM

This is a fantastic start, but I think it is missing an important piece of network overhead: replication overhead. As far as I can tell this is only injecting network costs for distsql senders. We also need to account for write traffic that is accounted to the raft group.

I suspect we historically did not account for the raft group network traffic, which was fine because it is represented in other cots (e.g. the mark up on each write and the replica storage cost). For tables that are widely replicated, e.g. global tables, the replication traffic is likely to make up the majority of the cross-region network traffic.

jaylim-crl committed:
tenantcostmodel: add RU multiplier for read and write requests

Previously, we multiplied consumed RU for write requests with the number of
replicas (i.e. replication factor). This commit decouples that, and introduces
new writeMultiplier (and readMultiplier) variables that explicitly denote
the multipliers for the write (and read) requests. For now, we will set
writeMultiplier=numReplicas and readMultiplier=1, and it is clear that this
model does not account for cross-region transfers since we are assuming that
all write requests (including replication) consume the same amount of RUs.
A future commit will address that.

This commit should be purely mechanical, and is a no-op.

Epic: CRDB-19074

Release note: None

jaylim-crl reviewed (commented):
TFTRs!

re: @JeffSwenson's comment. We've discussed offline that we are going with a different approach. The original approach of using the DistSender node as the source and the replica's node as the destination isn't accurate for writes since writes are then replicated from and to different nodes. The PR has been updated to take into account all the localities of the replicas, and we'd remove numReplicas from the multiplier for write requests.

PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-:-NN8btj2AwK9S7MLCDtr:btr1kkp)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @RaduBerinde)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2245 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMq7Bpt61VrCYE4nCu1:-NNKQlDkB7qPA7uetW3R:b-7xisri) ([raw file](https://github.com/cockroachdb/cockroach/blob/858dbeb96ff0b514bfc203d6c892a4bc8cec4b9c/pkg/kv/kvclient/kvcoord/dist_sender.go#L2245)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think this chunk of logic belongs in a helper function. Especially once you add some extra logging/comments, it deserves it's own function - it might even deserve a little unit test.
</blockquote></details>

Done. See `computeReadAndWriteRUMultipliers`.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2247 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMq6hMe7VHNW1CwXQwg:-NN8QNFvB1nKbigO1nz-:b-40bqd8) ([raw file](https://github.com/cockroachdb/cockroach/blob/858dbeb96ff0b514bfc203d6c892a4bc8cec4b9c/pkg/kv/kvclient/kvcoord/dist_sender.go#L2247)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: It's unfortunate that we hardcode a particular locality tier name here. Ideally, we'd have a cluster setting that specifies the name or some other way to configure it. That said, it's possible and maybe likely it'd ever be used for anything but Serverless, so not sure it's worth the trouble. At the very least, we should have a comment about it.
</blockquote></details>

Done. Note that this applies to the entire codebase, and not just the code here. There are other places that assume the `region` key.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMq7LPv2WBuGR77cpsi:-NN8brGX-4V0Ye05rASi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/858dbeb96ff0b514bfc203d6c892a4bc8cec4b9c/pkg/kv/kvclient/kvcoord/dist_sender.go#L2248)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Similar to below, I think we need to log an error if we can't find "region" in the tiers.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/replica_rate_limit.go` line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMq7wpSCs-BrcP-A0tZ:-NN8P9KM-YVKnQ4EZV0b:b-n0alok) ([raw file](https://github.com/cockroachdb/cockroach/blob/858dbeb96ff0b514bfc203d6c892a4bc8cec4b9c/pkg/kv/kvserver/replica_rate_limit.go#L31)):*
I was referring to the two fields originally - fromRegion and toRegion. I clarified the comment:
> {write,read}Multiplier isn't needed here since it's only used to calculate RUs.
___
*[`pkg/multitenant/tenantcostmodel/model.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMq8sXi5knxPeBwm1Rw:-NN8PPl08FmjrW6VeC48:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/858dbeb96ff0b514bfc203d6c892a4bc8cec4b9c/pkg/multitenant/tenantcostmodel/model.go#L18)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Do you mind updating this comment to read:
```
To get an idea of the magnitude of an RU, the cost model was designed so that a small point read of < 64 bytes costs 1 RU.
```
</blockquote></details>

Done.
___
*[`pkg/multitenant/tenantcostmodel/settings.go` line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMgl0r6TAq6mkupk1KF-r1-168:-NN8PVzaAPkdsNvJtw3y:b-evfgth) ([raw file](https://github.com/cockroachdb/cockroach/blob/858dbeb96ff0b514bfc203d6c892a4bc8cec4b9c/pkg/multitenant/tenantcostmodel/settings.go#L168)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

nit: I think we should use snake_case for the json keys.

CRDB isn't completely consistent with json style, but greping through the code base suggests that snake case is the dominant json naming convention.

The only other setting that accepts json is https://github.com/cockroachdb/cockroach/blob/fefa5fe16cc38c14a1406f6c91002493faa1cfb4/pkg/ccl/oidcccl/settings.go#L104 and it uses snake case as well.
</blockquote></details>

Done. Note that in our case, it doesn't matter if we specify a key since `json.Unmarshal` accepts a case-insensitive match. 

I also see that `changefeed.node_throttle_config` does not use snake case:
https://github.com/cockroachdb/cockroach/blob/8ae413bd7157fbaf11885772aa694671ace717b5/pkg/ccl/changefeedccl/changefeedbase/settings.go#L135-L150

Regardless, I updated this accordingly.
___
*[`pkg/multitenant/tenantcostmodel/settings.go` line 278 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMqoUXJDfIAz9hYldLn:-NN8caGX8DIKxC76ji57:b-qi6piu) ([raw file](https://github.com/cockroachdb/cockroach/blob/858dbeb96ff0b514bfc203d6c892a4bc8cec4b9c/pkg/multitenant/tenantcostmodel/settings.go#L278)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

How come we don't skip adding `1` multipliers into this table, since we always fall back to `1` during lookup if we can't find region(s)?
</blockquote></details>

I wanted to make sure the table represents what the user sets as a cluster setting, which in this case, a cluster with 3 regions should point to a 3x3 matrix.
___
*[`pkg/multitenant/tenantcostmodel/settings.go` line 290 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMqnQo-8ngJD4vPsfBk:-NN8cNGaBJOqavWW8aTx:b-yup62i) ([raw file](https://github.com/cockroachdb/cockroach/blob/858dbeb96ff0b514bfc203d6c892a4bc8cec4b9c/pkg/multitenant/tenantcostmodel/settings.go#L290)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: Did you consider only putting the 1-way mappings in this table and checking both sides during lookup? It'd mean 2 lookups 50% of the time and 3 lookups 50% of the time, vs. 2 lookups 100% of the time. But it'd use 1/2 the memory. I think this way is fine, but just wanted to check that you'd made a conscious decision to go this way.
</blockquote></details>

Yes, that was my initial approach. I realized it makes the code much subtle. Doing it this way could also help in debugging in the future, if one decides to print the entire matrix.


<!-- Sent from Reviewable.io -->

JeffSwenson commented on pkg/multitenant/tenantcostmodel/model.go:
```diff
@@ -111,38 +120,46 @@ func (c *Config) ExternalIOIngressCost(bytes int64) RU {
 	cost += RU(bri.writeCount) * c.KVWriteRequest
 	cost += RU(bri.writeBytes) * c.KVWriteByte
-	return cost * RU(bri.writeReplicas)
+	return cost * RU(bri.writeMultiplier)
```
cc: @andy-kimball @jaylim-crl 

There is a design question I think we should discuss. This is applying the writeMultiplier to the request overhead and the network overhead. An alternative approach is we would use a flat count(replicas) multiplier for the request overhead and the variable multiplier for the network overhead.

Most of the cost we are trying to account for is the intra vs inter region networking costs. The CPU cost of a local operation is similar to the CPU cost of a cross-region operation.

I'm in favor of only applying the variable multiplier to the network cost, but I think there is a valid case to be made for applying the cost table to both.

There are some additional costs for the cross region write that would be nice to account for. The cross region write has to be buffered in memory for a longer period of time and is more likely to require retries.

JeffSwenson commented on pkg/kv/kvclient/kvcoord/dist_sender_test.go:
```diff
@@ -5130,6 +5134,188 @@ func TestSendToReplicasSkipsStaleReplicas(t *testing.T) {
+	// Set regional cost multiplier table.
+	//                     | us-east1 | eu-central1 | asia-southeast1
+	//     -----------------------------------------------------------
+	//        us-east1     |    1     |     1.5     |        2
```
nit: the table in this comment does not match the table in the configuration.

(us-east, asia-southeast1) should be 2.6 instead of 2.

JeffSwenson commented on pkg/kv/kvclient/kvcoord/dist_sender_test.go:
It would be nice to include a few tests cases for complex, but expected scenarios. Here are a cases I think we should cover recommendations:

### Simulate a Global Table Local
client: us-east1
currentReplica: us-east1
replicas: {"us-east1": 3, "eu-central1": 1, "asia-southeast1": 1}
expectedWriteMultipler: (3 * 1) + (1 * 1.5) + (1 * 2.6) = 7.1
expectedReadMultipler: 1

### Write to Global Table from non-primary region
client: asia-southeast1
currentReplica: us-east1
replicas: {"us-east1": 3, "eu-central1": 1, "asia-southeast1": 1}
expectedWriteMultipler: (3 * 1.5) + (1 * 1) + (1 * 3.5) = 12.3
expectedReadMultipler: 3.5

### Read from to Global Table in non-primary region
client: eu-central1
currentReplica: eu-central1
replicas: {"us-east1": 3, "eu-central1": 1, "asia-southeast1": 1}
expectedWriteMultipler: (3 * 2.6) + (1 * 3.5) + (1 * 1) = 9
expectedReadMultipler: 1

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95913)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jaylim-crl, @JeffSwenson, and @RaduBerinde)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2389 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNNX96h6GlecKLQKdVu:-NNNX96i3U7lDhOIO9YB:b-bp9cvk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b366f8cc546a5f7f899d98ab073dd9030e20a6cf/pkg/kv/kvclient/kvcoord/dist_sender.go#L2389)):*
> ```Go
> //
> // NOTE: desc cannot be nil.
> func (ds *DistSender) computeReadAndWriteRUMultipliers(
> ```

It's not ideal that we're computing the write multiplier even when all we need is the read multiplier. Before, it was trivial to do both, but now we're performing an allocation and iterating over all replicas, even in cases where we don't need the write multiplier. What do you think about splitting computation of read/write multipliers into separate functions, and then only invoking one or the other. If you look at `MakeRequestInfo` and `MakeResponseInfo`, you'll see that they immediately short-circuit and return an empty struct based on whether it's a read or write op.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2412 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNNJDBB46vTp8c0ORNB:-NNNJDBB46vTp8c0ORNC:b8mwa68) ([raw file](https://github.com/cockroachdb/cockroach/blob/b366f8cc546a5f7f899d98ab073dd9030e20a6cf/pkg/kv/kvclient/kvcoord/dist_sender.go#L2412)):*
> ```Go
> 	// A leaseholder is not needed here since we are interested in returning
> 	// all replicas, which will already include the leaseholder by definition.
> 	replicas, err := NewReplicaSlice(ctx, ds.nodeDescs, desc, nil /* leaseholder */, AllReplicas)
> ```

This will perform at least one allocation. Is it the only way to get what you need, or is there some other way to iterate over replicas without creating a slice?
___
*[`pkg/multitenant/tenantcostmodel/model.go` line 141 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMgl0r6TAq6mkupk1KE-r3-141:-NNNGa1924MdrRwa7aEZ:b-l1be67) ([raw file](https://github.com/cockroachdb/cockroach/blob/b366f8cc546a5f7f899d98ab073dd9030e20a6cf/pkg/multitenant/tenantcostmodel/model.go#L141)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

cc: @andy-kimball @jaylim-crl 

There is a design question I think we should discuss. This is applying the writeMultiplier to the request overhead and the network overhead. An alternative approach is we would use a flat count(replicas) multiplier for the request overhead and the variable multiplier for the network overhead.

Most of the cost we are trying to account for is the intra vs inter region networking costs. The CPU cost of a local operation is similar to the CPU cost of a cross-region operation.

I'm in favor of only applying the variable multiplier to the network cost, but I think there is a valid case to be made for applying the cost table to both.

There are some additional costs for the cross region write that would be nice to account for. The cross region write has to be buffered in memory for a longer period of time and is more likely to require retries.

</blockquote></details>

I think that applying the multiplier to the total cost (which includes both batch/request & network costs) is preferable, since it's easier to understand and explain. That's much more important than precision, since this formula is an imprecise estimate to begin with.
___
*[`pkg/multitenant/tenantcostmodel/model.go` line 166 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNNIOj3C9fZ7eSnk_ok:-NNNIOj3C9fZ7eSnk_ol:b-2qvqhg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b366f8cc546a5f7f899d98ab073dd9030e20a6cf/pkg/multitenant/tenantcostmodel/model.go#L166)):*
> ```Go
> 	// writeReplicas is the number of range replicas to which this write was sent
> 	// (i.e. the replication factor). This is 0 if it is a read-only batch.
> 	writeReplicas int64
> ```

Doesn't your new code mean we can get rid of this, since the `writeMultiplier` now incorporates the # replicas that we're writing to?


<!-- Sent from Reviewable.io -->

jaylim-crl reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-:-NNNpXyzBLXuxMXbGGuH:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @JeffSwenson, and @RaduBerinde)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2389 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNNX96h6GlecKLQKdVu:-NNNrDR89k4V2kdZjhJN:b-gsmqrt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b366f8cc546a5f7f899d98ab073dd9030e20a6cf/pkg/kv/kvclient/kvcoord/dist_sender.go#L2389)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

It's not ideal that we're computing the write multiplier even when all we need is the read multiplier. Before, it was trivial to do both, but now we're performing an allocation and iterating over all replicas, even in cases where we don't need the write multiplier. What do you think about splitting computation of read/write multipliers into separate functions, and then only invoking one or the other. If you look at `MakeRequestInfo` and `MakeResponseInfo`, you'll see that they immediately short-circuit and return an empty struct based on whether it's a read or write op.
</blockquote></details>

I can split into two functions, and only call one or the other if necessary. However, we will only know if it is a write request after processing `MakeRequestInfo`: https://github.com/cockroachdb/cockroach/blob/899c9112188c22ec140adbeb0cf7d5761da3827d/pkg/kv/kvclient/kvcoord/dist_sender.go#L2246-L2247

How do you feel about either of the following?
1. Introduce a new `SetMultiplier` function that gets called after a request/response info was made.
2. Pass in a callback into MakeRequestInfo or MakeResponseInfo which gets called only if we don't short-circuit.

I have a feeling @JeffSwenson may object to (1) based on what I've been doing to the proxy back then.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2412 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNNJDBB46vTp8c0ORNB:-NNNq-aREEnGFCmkPs8Z:b-ye26gu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b366f8cc546a5f7f899d98ab073dd9030e20a6cf/pkg/kv/kvclient/kvcoord/dist_sender.go#L2412)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This will perform at least one allocation. Is it the only way to get what you need, or is there some other way to iterate over replicas without creating a slice?
</blockquote></details>

I see two alternatives:
1. Extract the code in NewReplicaSlice, and call what we need directly, where `replicas=desc.Replicas().Descriptors()`: https://github.com/cockroachdb/cockroach/blob/899c9112188c22ec140adbeb0cf7d5761da3827d/pkg/kv/kvclient/kvcoord/replica_slice.go#L115-L128
2. DistSender already calls NewReplicaSlice each time we send a batch: https://github.com/cockroachdb/cockroach/blob/899c9112188c22ec140adbeb0cf7d5761da3827d/pkg/kv/kvclient/kvcoord/dist_sender.go#L2015-L2020
We can theoretically update the filter to return all, but this also means we're breaking the network hop optimization that we had for regular paths: https://github.com/cockroachdb/cockroach/blob/899c9112188c22ec140adbeb0cf7d5761da3827d/pkg/kv/kvclient/kvcoord/replica_slice.go#L89-L91

I could update the code to do (1) directly, if you'd prefer it that way. I was focusing on simplicity earlier.
___
*[`pkg/multitenant/tenantcostmodel/model.go` line 166 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNNIOj3C9fZ7eSnk_ok:-NNNpa5Y4Gi_oo_8iFVN:b-yykj9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/b366f8cc546a5f7f899d98ab073dd9030e20a6cf/pkg/multitenant/tenantcostmodel/model.go#L166)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Doesn't your new code mean we can get rid of this, since the `writeMultiplier` now incorporates the # replicas that we're writing to?
</blockquote></details>

That was the original thought, but it appears that we can't do that because the number of replicas is used to compute the number of batches:

https://github.com/cockroachdb/cockroach/blob/899c9112188c22ec140adbeb0cf7d5761da3827d/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L805-L811


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95913)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jaylim-crl, @JeffSwenson, and @RaduBerinde)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2389 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNNX96h6GlecKLQKdVu:-NNP07g53QYEWg7igqXj:b-thqshw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b366f8cc546a5f7f899d98ab073dd9030e20a6cf/pkg/kv/kvclient/kvcoord/dist_sender.go#L2389)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

I can split into two functions, and only call one or the other if necessary. However, we will only know if it is a write request after processing `MakeRequestInfo`: https://github.com/cockroachdb/cockroach/blob/899c9112188c22ec140adbeb0cf7d5761da3827d/pkg/kv/kvclient/kvcoord/dist_sender.go#L2246-L2247

How do you feel about either of the following?
1. Introduce a new `SetMultiplier` function that gets called after a request/response info was made.
2. Pass in a callback into MakeRequestInfo or MakeResponseInfo which gets called only if we don't short-circuit.

I have a feeling @JeffSwenson may object to (1) based on what I've been doing to the proxy back then.
</blockquote></details>

I was thinking something like:
```
var reqInfo tenantcostmodel.RequestInfo
var respInfo tenantcostmodel.ResponseInfo
if ba.IsWrite() {
  reqInfo = tenantcostmodel.MakeRequestInfo(ba, ds.computeWriteRUMultiplier(ctx, desc))
} else {
  respInfo = tenantcostmodel.MakeResponseInfo(br, true, ds.computeReadRUMultiplier(ctx, &curReplica))
}
if err := ds.kvInterceptor.OnResponseWait(ctx, reqInfo, respInfo); err != nil {
  return nil, err
}
```
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2412 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNNJDBB46vTp8c0ORNB:-NNP1AzJ57rCpCnohHDJ:bpmpwdj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b366f8cc546a5f7f899d98ab073dd9030e20a6cf/pkg/kv/kvclient/kvcoord/dist_sender.go#L2412)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

I see two alternatives:
1. Extract the code in NewReplicaSlice, and call what we need directly, where `replicas=desc.Replicas().Descriptors()`: https://github.com/cockroachdb/cockroach/blob/899c9112188c22ec140adbeb0cf7d5761da3827d/pkg/kv/kvclient/kvcoord/replica_slice.go#L115-L128
2. DistSender already calls NewReplicaSlice each time we send a batch: https://github.com/cockroachdb/cockroach/blob/899c9112188c22ec140adbeb0cf7d5761da3827d/pkg/kv/kvclient/kvcoord/dist_sender.go#L2015-L2020
We can theoretically update the filter to return all, but this also means we're breaking the network hop optimization that we had for regular paths: https://github.com/cockroachdb/cockroach/blob/899c9112188c22ec140adbeb0cf7d5761da3827d/pkg/kv/kvclient/kvcoord/replica_slice.go#L89-L91

I could update the code to do (1) directly, if you'd prefer it that way. I was focusing on simplicity earlier.
</blockquote></details>

Well, my main concern is that we're doing this allocation in every case, even the common read path where we don't even need it. If you split multiplier calculation for read vs. write, then I think it'd be fine to just leave the code simple, since we'd then only pay in the write path.
___
*[`pkg/multitenant/tenantcostmodel/model.go` line 166 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNNIOj3C9fZ7eSnk_ok:-NNP1oDD2HayekZ9577j:b-j772f0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b366f8cc546a5f7f899d98ab073dd9030e20a6cf/pkg/multitenant/tenantcostmodel/model.go#L166)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

That was the original thought, but it appears that we can't do that because the number of replicas is used to compute the number of batches:

https://github.com/cockroachdb/cockroach/blob/899c9112188c22ec140adbeb0cf7d5761da3827d/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L805-L811
</blockquote></details>

Hm, I see, yeah, I don't see an obvious way around that.


<!-- Sent from Reviewable.io -->

jaylim-crl commented:
TFTR! 

Reviewable doesn't seem to be working for me now, so I'll comment directly.

> I was thinking something like:

```go
var reqInfo tenantcostmodel.RequestInfo
var respInfo tenantcostmodel.ResponseInfo
if ba.IsWrite() {
  reqInfo = tenantcostmodel.MakeRequestInfo(ba, ds.computeWriteRUMultiplier(ctx, desc))
} else {
  respInfo = tenantcostmodel.MakeResponseInfo(br, true, ds.computeReadRUMultiplier(ctx, &curReplica))
}
if err := ds.kvInterceptor.OnResponseWait(ctx, reqInfo, respInfo); err != nil {
  return nil, err
}
```

Hm, I see it now. There's a `KVWriteBatch` that includes the base cost, and we also want to multiply that. My original thought was to use `reqInfo.IsWrite()` instead of just `ba.IsWrite()`, but looks like that will not work. The idea that you proposed above sounds reasonable.

> Well, my main concern is that we're doing this allocation in every case, even the common read path where we don't even need it. If you split multiplier calculation for read vs. write, then I think it'd be fine to just leave the code simple, since we'd then only pay in the write path.

Sounds good - will split and leave it as-is.

jaylim-crl commented:
TFTRs! I believe I have addressed all the reviews. PTAL.

(Still having issue trying to use reviewable. I wonder if this is related to the recent GitHub infra changes)

> I was thinking something like: ...

Done.

> Well, my main concern is that we're doing this allocation in every case, even the common read path where we don't even need it. If you split multiplier calculation for read vs. write, then I think it'd be fine to just leave the code simple, since we'd then only pay in the write path.

Done. Given that the RU computation function has been separated into two now, I think this shouldn't be an issue anymore.

> nit: the table in this comment does not match the table in the configuration.
(us-east, asia-southeast1) should be 2.6 instead of 2.

Done. 

> It would be nice to include a few tests cases for complex, but expected scenarios. Here are a cases I think we should cover recommendations:

Done.

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, with a couple nits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-:-NNPxWZP4eB942dmRmq0:bef3uqw)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jaylim-crl, @JeffSwenson, and @RaduBerinde)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2479 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNPuT5x6P67-X2So-_F:-NNPuT5x6P67-X2So-_G:b8sa2hn) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b0e7881b1e8eafe3f66fe1c47a9fc2771d3a05/pkg/kv/kvclient/kvcoord/dist_sender.go#L2479)):*
> ```Go
> // computeReadRUMultiplier returns the read RU multiplier for a batch that is
> // sent from the current DistSender node to the node with curReplica.
> func (ds *DistSender) computeReadRUMultiplier(
> ```

NIT: I was hoping the read case would be much simpler, but I see now that you have to do pretty much the same stuff here as for the write path. It kills me a little bit inside to see so much code duplication. What about using some version of your previous code, but passing an `isRead` parameter to select whether to compute a read multiplier or a write multiplier? Or maybe there's a function you could factor out here? Not a blocker, though.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2511 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNPwRBH4z4PZTxb_XsB:-NNPwRBH4z4PZTxb_XsC:bnsj6o1) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b0e7881b1e8eafe3f66fe1c47a9fc2771d3a05/pkg/kv/kvclient/kvcoord/dist_sender.go#L2511)):*
> ```Go
> 
> 	var toRegion string
> 	for _, r := range replicas {
> ```

NIT: Can't you use `replicas.Find` to find the matching `ReplicaID`?


<!-- Sent from Reviewable.io -->

jaylim-crl commented:
TFTR!

> NIT: I was hoping the read case would be much simpler, but I see now that you have to do pretty much the same stuff here as for the write path. It kills me a little bit inside to see so much code duplication. What about using some version of your previous code, but passing an isRead parameter to select whether to compute a read multiplier or a write multiplier? Or maybe there's a function you could factor out here? Not a blocker, though.

Hm - I'll take a look. I can get it to conditionally compute either a read or write multiplier.

> NIT: Can't you use replicas.Find to find the matching ReplicaID?

I had that in the first version of this PR, but I feel like the complexity is similar - one needs to check for a -1 index, whereas the other loops over `replicas`. Regardless, will update to use `.Find`.

jaylim-crl reviewed (commented):
TFTRs! All comments have been addressed.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-:-NNZQ2d4BFdTTJPLU8Qm:bgjvd9j)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @JeffSwenson, and @RaduBerinde)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2389 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNNX96h6GlecKLQKdVu:-NNZPhim9LSL1blNFgd6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b366f8cc546a5f7f899d98ab073dd9030e20a6cf/pkg/kv/kvclient/kvcoord/dist_sender.go#L2389)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I was thinking something like:
```
var reqInfo tenantcostmodel.RequestInfo
var respInfo tenantcostmodel.ResponseInfo
if ba.IsWrite() {
  reqInfo = tenantcostmodel.MakeRequestInfo(ba, ds.computeWriteRUMultiplier(ctx, desc))
} else {
  respInfo = tenantcostmodel.MakeResponseInfo(br, true, ds.computeReadRUMultiplier(ctx, &curReplica))
}
if err := ds.kvInterceptor.OnResponseWait(ctx, reqInfo, respInfo); err != nil {
  return nil, err
}
```
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2479 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNPuT5x6P67-X2So-_F:-NNZPjY2Atg9LqafZA70:big6h0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b0e7881b1e8eafe3f66fe1c47a9fc2771d3a05/pkg/kv/kvclient/kvcoord/dist_sender.go#L2479)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: I was hoping the read case would be much simpler, but I see now that you have to do pretty much the same stuff here as for the write path. It kills me a little bit inside to see so much code duplication. What about using some version of your previous code, but passing an `isRead` parameter to select whether to compute a read multiplier or a write multiplier? Or maybe there's a function you could factor out here? Not a blocker, though.
</blockquote></details>

Done. I moved everything into a single `computeSendRUMultiplier` method, which I think is much cleaner now.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2511 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNPwRBH4z4PZTxb_XsB:-NNZPpM_5GVllNHk8lhl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b0e7881b1e8eafe3f66fe1c47a9fc2771d3a05/pkg/kv/kvclient/kvcoord/dist_sender.go#L2511)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: Can't you use `replicas.Find` to find the matching `ReplicaID`?
</blockquote></details>

Done.
___
*[`pkg/multitenant/tenantcostmodel/model.go` line 141 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMgl0r6TAq6mkupk1KE-r3-141:-NNZQ9eO8F_TnL6sW1Rq:b-xrkltj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b366f8cc546a5f7f899d98ab073dd9030e20a6cf/pkg/multitenant/tenantcostmodel/model.go#L141)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think that applying the multiplier to the total cost (which includes both batch/request & network costs) is preferable, since it's easier to understand and explain. That's much more important than precision, since this formula is an imprecise estimate to begin with.
</blockquote></details>

Discussed offline. I left this as-is based on the thoughts that we have so far (i.e. simplicity over preciseness). Happy to revisit this if this is still an issue.
___
*[`pkg/multitenant/tenantcostmodel/model.go` line 166 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNNIOj3C9fZ7eSnk_ok:-NNZPsss8yabUnZ7SdKN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b366f8cc546a5f7f899d98ab073dd9030e20a6cf/pkg/multitenant/tenantcostmodel/model.go#L166)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Hm, I see, yeah, I don't see an obvious way around that.
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 5159 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNKPbp3DDJiHPg4V9fR-r3-5159:-NNZPtte7V9I7yKWOXiG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b366f8cc546a5f7f899d98ab073dd9030e20a6cf/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L5159)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

nit: the table in this comment does not match the table in the configuration.

(us-east, asia-southeast1) should be 2.6 instead of 2.
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 5201 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NNKPbp3DDJiHPg4V9fR-r3-5201:-NNZPuSs1UeDMQwuy0jY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b366f8cc546a5f7f899d98ab073dd9030e20a6cf/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L5201)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

It would be nice to include a few tests cases for complex, but expected scenarios. Here are a cases I think we should cover recommendations:

### Simulate a Global Table Local
client: us-east1
currentReplica: us-east1
replicas: {"us-east1": 3, "eu-central1": 1, "asia-southeast1": 1}
expectedWriteMultipler: (3 * 1) + (1 * 1.5) + (1 * 2.6) = 7.1
expectedReadMultipler: 1

### Write to Global Table from non-primary region
client: asia-southeast1
currentReplica: us-east1
replicas: {"us-east1": 3, "eu-central1": 1, "asia-southeast1": 1}
expectedWriteMultipler: (3 * 1.5) + (1 * 1) + (1 * 3.5) = 12.3
expectedReadMultipler: 3.5

### Read from to Global Table in non-primary region
client: eu-central1
currentReplica: eu-central1
replicas: {"us-east1": 3, "eu-central1": 1, "asia-southeast1": 1}
expectedWriteMultipler: (3 * 2.6) + (1 * 3.5) + (1 * 1) = 9
expectedReadMultipler: 1
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

JeffSwenson commented on pkg/kv/kvclient/kvcoord/dist_sender.go:
```diff
@@ -2242,9 +2242,24 @@ func (ds *DistSender) sendToReplicas(
-					numReplicas := len(desc.Replicas().Descriptors())
-					reqInfo := tenantcostmodel.MakeRequestInfo(ba, numReplicas)
-					respInfo := tenantcostmodel.MakeResponseInfo(br, !reqInfo.IsWrite())
+					var reqInfo tenantcostmodel.RequestInfo
```
I liked your original implementation better. I don't think splitting the read and write paths saved an allocation. reqInfo and respInfo are flat structs with no embedded pointers.

jeffswenson reviewed (approved):
LGTM

jaylim-crl reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-:-NNdFDDy5wsHYg193sfA:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @JeffSwenson, and @RaduBerinde)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2245 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMgl0r6TAq6mkupk1KA-r6-2245:-NNdEsOEE3kDP2javMHS:b-q882ul) ([raw file](https://github.com/cockroachdb/cockroach/blob/b61a1d33194753121287ea4a129173b9e54df4b5/pkg/kv/kvclient/kvcoord/dist_sender.go#L2245)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

I liked your original implementation better. I don't think splitting the read and write paths saved an allocation. reqInfo and respInfo are flat structs with no embedded pointers.
</blockquote></details>

It's `NewReplicaSlice` that is allocating. The write path invokes that. If we didn't split, there's a possibility where `ba.IsWrite()` is false, but we're still performing an unnecessary allocation. 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2256 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/95913#-NMgl0r6TAq6mkupk1KA-r6-2256:-NNdEL5cC5AL8YR3hu6W:b-qdgwcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b61a1d33194753121287ea4a129173b9e54df4b5/pkg/kv/kvclient/kvcoord/dist_sender.go#L2256)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

Why is the second branch if `!reqInfo.IsWrite()` instead of an else block? E.g. 

```
if ba.IsWrite() {
  // initialize MakeRequestInfo
} else {
  // initialize MakeResponseInfo
}

```
</blockquote></details>

I wanted to maintain the existing behavior: 
https://github.com/cockroachdb/cockroach/blob/eb158026c50d8fa856e42f928d844831ea9e6b28/pkg/kv/kvclient/kvcoord/dist_sender.go#L2245-L2250

Notice how we pass in `reqInfo.IsWrite()` when constructing the response info, which indicated that there is a possibility where both `ba.IsWrite()` and `!reqInfo.IsWrite()` are true, especially when the batch contains all admin requests:

https://github.com/cockroachdb/cockroach/blob/eb158026c50d8fa856e42f928d844831ea9e6b28/pkg/multitenant/tenantcostmodel/model.go#L166-L180


<!-- Sent from Reviewable.io -->

jaylim-crl commented:
TFTRs!

bors r=JeffSwenson,andy-kimball
